python/desc/imsim/tree_rings.py

```diff
@@ -0,0 +1,90 @@
+"""
+Code to add tree ring info to the CCD detectors
+"""
+import numpy as np
+import galsim
+
+__all__ = ['TreeRings']
+
+class TreeRingsError(Exception):
+    pass
+
+class TreeRings():
+    """
+    # Craig Lage UC Davis 16-Mar-18; cslage@ucdavis.edu
+    # This function returns a tree ring model drawn from an analytical function that was
+    # derived based on tree ring data collected by Hye-Yun Park at BNL.  The data
+    # used is in imSim/data/tree_ring_data, and a description of the method is in
+    # imSim/data/tree_ring_data/Tree_Rings_13Feb18.pdf
+    # Based on the data, 40% of the sensors are assumed to have 'bad' tree rings, with an
+    # amplitude 10X greater that the 60% of the sensors that have 'good' tree rings.
+    """
+
+    def __init__(self, tr_filename):
+        """
+        Constructor.
+        Craig Lage UC Davis 19-Mar-18; cslage@ucdavis.edu
+        This code reads in a file with tree ring parameters from tr_filename
+        and assigns a tree ring model to each sensor.
+        """
+        self.numfreqs = 20 # Number of spatial frequencies
+        self.cfreqs = np.zeros([self.numfreqs]) # Cosine frequencies
+        self.cphases = np.zeros([self.numfreqs])
+        self.sfreqs = np.zeros([self.numfreqs]) # Sine frequencies
+        self.sphases = np.zeros([self.numfreqs])
+        self.A = 0.0
+        self.B = 0.0
+        self.r_max = 8000.0 # Maximum extent of tree ring function in pixels
+        dr = 3.0 # Step size of tree ring function in pixels
+        self.npoints = int(self.r_max / dr) + 1 # Number of points in tree ring function
+
+        with open(tr_filename, 'r') as input_:
+            self.lines = input_.readlines() # Contents of tree_ring_parameters file
+
+        return
+    
+    def Read_DC2_Tree_Ring_Model(self, Rx, Ry, Sx, Sy):
+        """
+        This function finds the tree ring parameters for a given sensor
+        and assigns a tree ring model to that sensor.
+        """
+        try:
+            for i, line in enumerate(self.lines):
+                if line.split()[0] == 'Rx':
```

Comment:
Would it be more efficient to do the conversion to `int`s and `float`s once in the constructor, rather than storing the lines of the treering data file as strings and doing the conversion every time we call this method? 

Feel free to point out that this could only ever be called 189 times, so it's probably not worth the work to change it.

Reply:
I'm going to leave this one be.